### ISSUE TITLE
Allow tests to access real (non-mock) environment

### ISSUE THREAD

**Issue by jdegoes** [zio/zio#1474]
Currently tests and test aspects (like `timeout`) will always access the environment provided to them, which will be a default mock environment in `DefaultRunnableSpec`.

This is a serious obstacle, as test combinators need to use real timeouts, real system clocks, real consoles, etc. To solve the problem, inside `zio.test.mock` package, we can introduce a new module and service:

```scala
package zio.test.mock

trait Live[+R] {
  def live: Live.Service[R]
}
object Live {
  trait Service[+R] {
    def provide[E, A](zio: ZIO[R, E, A]): IO[E, A]
  }
  def live[R, E, A](zio: ZIO[R, E, A]): ZIO[Live[R], E, A] = 
    ZIO.accessM[Live[R]](_.live.provide(zio))
}
```

Now all test combinators (for example, `TestAspect.timeout`) can move from requiring, e.g. `Clock`, to requiring `Live[Clock]`.

Then the default environment that we provide in tests (`zio.test.mock.mockEnvironmentManaged` / `zio.test.MockEnvironment.Value`) can provide `Live[DefaultRuntime#Environment]`, by supplying an implementation of the live environment module that is backed by a "live" environment.

This should provide a clean way to get access to the "real" environment when necessary—as it is necessary for all test machinery—while not increasing number of type parameters or adding any complexity for the common case.

/cc @adamgfraser

**Comment by adamgfraser** [comment]
@jdegoes Sounds like a great idea. I will take this.

**Comment by ghostdogpr** [comment]
Btw shouldn’t tests have a timeout by default?

**Comment by adamgfraser** [comment]
@ghostdogpr I think so. I will take care of that as part of this.

**Comment by jdegoes** [comment]
@ghostdogpr @adamgfraser 

I do not think "timeout by default" makes sense. A user will have a long-running test and be surprised when it is killed, and spend a long time to find the source of the problem (the test framework did something "magical" to be helpful but it just ended up being surprising).

Worse, compositional timeouts always "narrow". They cannot "widen". This means if you have a default, i.e. if you apply `timeout(60.seconds) { ... }` to every test automatically, then that becomes the new upper bound on the time a test can execute.

Instead of default timeouts, I'd suggest that `TestRunner` could be a place to include a default timeout for a _whole test suite_, as a default parameter, perhaps; or _even better_, a test runner can  print out warning messages after X minutes without any test progress. 

The warning message could say something quite helpful, such as:

> A ZIO Spec has taken more than 5 minutes to execute. If this is not expected, consider using `TestAspect.timeout` to timeout runaway tests for faster diagnostics.
> ...
> A ZIO Spec has taken more than 10 minutes to execute. If this is not expected, consider using `TestAspect.timeout` to timeout runaway tests for faster diagnostics.
> ...

This will be enough to solve both "no output" and "no surprise" problems.

**Comment by adamgfraser** [comment]
@jdegoes That makes sense. I think the lack of compositionality is a very good point. A lot of times it is the lack of any output that is the real issue so If we can display a useful message that something is going on but a test is taking a long time that is great. It would be really nice if we could display as part of that which test or tests are taking a long time.

**Comment by jdegoes** [comment]
@adamgfraser Yes, the paths to the test are known by the `TestRunner`, basically a non empty `List[L]`, so we could display something more detailed, like:

> `In Suite "Integers", test "addition" has taken more than ....

**Comment by adamgfraser** [comment]
Okay, let’s get a couple of these other outstanding PRs in and then I can work on adding that.

**Comment by ghostdogpr** [comment]
The reason I suggested this is that when I was fixing a flaky test, the whole test suite got stuck with no output and I had to kill sbt to get out of it after a few minutes, without even knowing which test caused this. Not the best experience. 

Adding some output would at least be helpful to identify the faulty test, even though it wouldn’t run other tests. But then what about CI? If a flaky test causes the whole thing to be stuck forever, we’re going to have stuck jobs.
